Fix `gedcom_view.py` for very large GEDCOM file.

changeling opened the pull request:
Fix `gedcom_view.py` for very large GEDCOM files as input.

This should correct the python issue with handling `read()` for files in the multi-gigabyte range in `open()` in `gedcom_view.py`.

briot commented on gedcom_view.py:
```diff
@@ -24,13 +24,14 @@ def green(s):
 else:
-    gedcom = open(sys.argv[1], "r")
+    with open(sys.argv[1], 'r', buffering=1) as gedcom:
+        buffer = gedcom.readlines()
```
This is still allocating a lot of memory, since we are storing the whole file, albeit in a better way maybe.
I think we could do without `buffer` here, and directly use `for line in gedcom` below. 
Some long term plans is to use `mmap` to read the file, parse it to find the start of top-level (level 0) records. We then need to parse them in the right order (sources first, if I remember right), so that we can then parse the others, without having to keep the whole contents of the file in memory. Currently, we need to store `buffer` in memory, and then the result of splitting the lines, so basically we have the whole file twice in memory...

changeling replied:
Absolutely. This fix is merely intended to allow much larger files to actually be read. It doesn't address that they _are_ much larger files and eat a lot of memory. As far as those longer term plans, perhaps `diskcache` might be worth considering.

Also, I've removed the `buffer` usage, and now read directly from stdin or use `readlines()` to bring in the file, with `gedcom`.

changeling replied:
As a possible tool, if you aren't familiar with it:

http://www.grantjenks.com/docs/diskcache/tutorial.html

briot commented on gedcom_view.py:
```diff
@@ -24,13 +24,14 @@ def green(s):
 
 if len(sys.argv) == 1:
     gedcom = sys.stdin
+    buffer = gedcom.read().replace('\r\n', '\n').replace('\r', '\n')
```
Perhaps we no longer need the `replace` here... I had done that years ago, but python3 and `readlines` are in fact likely to already do this normalization

changeling replied:
Yes. Since v 3.0, python defaults to `Universal Newline` unless otherwise specified. I left that change until I could read up on stdin behavior, but have now removed the `replace`.

changeling commented:
Also, note that I've moved the colorizing into the `colorize()` function, and updated the output formatting to the new style.

One question: Is there a reason for using `sys.stdout.write()` rather than `print()`? `print()` defaults to `sys.stdout` unless the `file` value is specified. I've left it as `sys.stdout.write()` for now pending your thoughts.

changeling commented:
I've also removed the `line_num` counter pattern in favor of creating `line_number` along `line` using `enumerate(gedcom)`.

briot commented:
Thanks for the fixes !

changeling commented:
Glad I could contribute!

Cheers.

On Mon, Feb 18, 2019 at 12:51 AM Emmanuel Briot <notifications@github.com>
wrote:

> Thanks for the fixes !
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/briot/geneapro/pull/46#issuecomment-464607696>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAYd_sO4YCpUY-nDuPOkYmIcPo5_c71-ks5vOk2DgaJpZM4a_YrL>
> .
>
